### ISSUE TITLE
Different behavior in helm 2 and helm 3 during helm upgrade

### ISSUE THREAD

**Issue by shubhambhattar** [helm/helm#7058]
Output of `helm version`: 

```
version.BuildInfo{Version:"v3.0.0", GitCommit:"e29ce2a54e96cd02ccfce88bee4f58bb6e2a28b6", GitTreeState:"clean", GoVersion:"go1.13.4"}
```

Output of `kubectl version`: 

```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.8", GitCommit:"211047e9a1922595eaa3a1127ed365e9299a6c23", GitTreeState:"clean", BuildDate:"2019-10-15T12:11:03Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.5", GitCommit:"2166946f41b36dea2c4626f90a77706f426cdea2", GitTreeState:"clean", BuildDate:"2019-03-25T15:19:22Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.): AKS

---

I've noticed a breaking change (probably?) while migrating my charts from helm 2 to helm 3. 

The issue can be reproduced as follows:

I have a helm chart with the following structure (created with `helm3 create test`):

```
$ tree test/
test/
├── Chart.yaml
├── charts
├── templates
│   ├── Service.yaml
│   ├── _baseHelpers.tpl
│   └── _helpers.tpl
└── values.yaml

2 directories, 5 files

```

Contents of `_baseHelpers.tpl`:
```

$ cat test/templates/_baseHelpers.tpl
{{- define "base.labels" -}}
app.kubernetes.io/name: {{ include "test.name" . }}
helm.sh/chart: {{ include "test.chart" . }}
app.kubernetes.io/instance: {{ .Release.Name }}
{{- if .Chart.AppVersion }}
app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
{{- end }}
app.kubernetes.io/managed-by: {{ .Release.Service }}
{{- end -}}

```

Contents of `Service.yaml`:
```

cat test/templates/Service.yaml
{{- if .Values.Service }}

{{- $release_namespace := .Release.namespace }}
{{- $labels := include "base.labels" . }}

{{- range $service := .Values.Service }}
{{- with $service }}

{{- $type := default "ClusterIP" .type }}

apiVersion: v1
kind: Service
metadata:
  name: {{ .name }}
  namespace: {{ $release_namespace }}

  {{- with .annotations }}
  annotations:
    {{- toYaml . | nindent 4 }}
  {{- end }}
  labels:
{{ $labels | indent 4 }}
  {{- with .labels }}
    {{- toYaml . | nindent 4 }}
  {{- end }}

spec:
  {{- if or (eq $type "ClusterIP") (eq $type "NodePort") (eq $type "LoadBalancer") }}
  selector:
{{ $labels | indent 4 }}
    {{- with .selector }}
      {{- toYaml . | nindent 4 }}
    {{- end }}
  {{- end }}
  type: {{ $type | quote }}
  publishNotReadyAddresses: {{ default false .publishNotReadyAddresses }}
  sessionAffinity: {{ .sessionAffinity | default "None" | quote }}

  {{- if (eq $type "ClusterIP") }}
  clusterIP: {{ default nil .clusterIP | quote }}

  {{- else if (eq $type "ExternalName") }}
  externalName: {{ required "externalName is required in type ExternalName" .externalName | quote }}

  {{- else if (eq $type "NodePort") }}
  clusterIP: {{ default nil .clusterIP | quote }}
  externalTrafficPolicy: {{ .externalTrafficPolicy | default "Cluster" | quote }}

  {{- else if (eq $type "LoadBalancer") }}
  clusterIP: {{ default nil .clusterIP | quote }}
  loadBalancerIP: {{ default nil .loadBalancerIP | quote }}
  externalTrafficPolicy: {{ .externalTrafficPolicy | default "Cluster" | quote }}

  {{- with .loadBalancerSourceRanges }}
  loadBalancerSourceRanges:
    {{- toYaml . | nindent 4 }}
  {{- end }}

  {{- if (.externalTrafficPolicy) and (eq .externalTrafficPolicy "Local") }}
  healthCheckNodePort: {{ default nil .healthCheckNodePort | quote }}
  {{- end }}

  {{- end }}

  {{- with .ports }}
  ports:
    {{- toYaml . | nindent 4 }}
  {{- end }}

  {{- with .externalIPs }}
  externalIPs:
    {{- toYaml . | nindent 4 }}
  {{- end }}

  {{- with .sessionAffinityConfig }}
  sessionAffinityConfig:
    {{- toYaml . | nindent 4 }}
  {{- end }}
---
{{- end }}
{{- end }}
{{- end }}

```

Contents of `values.yaml`:
```

Service: [
  {
    name: test,
    type: ClusterIP,
    ports: [
      {
        name: http,
        protocol: TCP,
        port: 80,
        targetPort: 8080
      }
    ]
  }
]

```
Output of `helm3 template test`:
```

$ helm3 template test/
---
# Source: test/templates/Service.yaml
apiVersion: v1
kind: Service
metadata:
  name: test
  namespace:
  labels:
    app.kubernetes.io/name: test
    helm.sh/chart: test-0.1.0
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/version: "1.16.0"
    app.kubernetes.io/managed-by: Helm

spec:
  selector:
    app.kubernetes.io/name: test
    helm.sh/chart: test-0.1.0
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/version: "1.16.0"
    app.kubernetes.io/managed-by: Helm
  type: "ClusterIP"
  publishNotReadyAddresses: false
  sessionAffinity: "None"
  clusterIP:
  ports:
    - name: http
      port: 80
      protocol: TCP
      targetPort: 8080

```

If I install this chart for the first time, the chart is installed and no errors are reported:
```

$ helm3 install test test/
NAME: test
LAST DEPLOYED: Sat Nov 23 03:47:50 2019
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None

```

However, if I just change the `targetPort` in `values.yaml` to `8081` (or any other value) and try to upgrade the chart, I encounter an error:

```

$ helm3 upgrade test test/
Error: UPGRADE FAILED: cannot patch "test" with kind Service: Service "test" is invalid: spec.clusterIP: Invalid value: "": field is immutable

```

This behavior was not seen in helm 2. 

Is this an intended change? I want to migrate the charts in my team from helm 2 to helm 3 and I've created a few templates for everyone's use (this is what you see in `_baseHelpers.tpl` and `Service.yaml` - our custom templates).

The reason you see `clusterIP: ` in `helm3 template ...` output is because our templates account for all the fields that are possible in a Kubernetes Resource (Service, in this case). All the fields will be printed in `helm3 template ...` output and if they have a default value, then it will be automatically applied (ex: `publishNotReadyAddresses: false` or can be overriden in `values.yaml` file) and the optional fields will be rendered as `nil`. Please take a look at `Service.yaml` files' contents above.

**Comment by bacongobbler** [comment]
related: https://github.com/helm/helm/issues/6378

**Comment by bacongobbler** [comment]
tl;dr yes, this is expected behaviour and it was an intentional change. There are edge cases like service clusterIPs where the live state and the proposed state cause invalid patches to be generated. Omitting the `clusterIP:` definition on the service when you first install the chart should solve the issue, as one will be generated for you.

e.g. change this:

```
clusterIP: {{ default nil .clusterIP | quote }}
```

to an `if` clause that omits that field from the manifest if it hasn't been set will fix the issue.

In other words, don't do this:

```
clusterIP: ""
```

Or else on an upgrade, helm will try to change the service's clusterIP from an auto-generated IP address to the empty string, hence the error message.

**Comment by bacongobbler** [comment]
Closing as a duplicate of #6378 so we can keep the conversation in one place. Please also see the detailed explanation and suggested course of action in https://github.com/helm/helm/issues/6378#issuecomment-557746499. Thanks!

**Comment by shubhambhattar** [comment]
@bacongobbler Thanks for the explanation and the related issue. For anyone else looking at this issue, [this comment](https://github.com/helm/helm/issues/6378#issuecomment-557746499) might be helpful.
